sdk/identity/Azure.Identity/src/EnvironmentCredential.cs

```diff
@@ -206,5 +165,144 @@ private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestConte
+        /// will non null.
+        /// </summary>
+        /// <remarks>
+        /// AZURE_AUTH_LOCATION has a lower precedence than the other AZURE_* environment variaibles. When a complete set of these variaibles are set
+        /// (i.e. AZURE_TENANT_ID, AZURE_CLIENT_ID and either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD) they are prefered even if
+        /// AZURE_AUTH_LOCATON is set.
+        /// </remarks>
+        /// <param name="isAsync">When true, this method may complete asynchronously.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling cancelation.</param>
+        /// <returns></returns>
+        private async ValueTask EnsureInitialized(bool isAsync, CancellationToken cancellationToken)
+        {
+            // If we've already tried to initalize the credential and failed, don't try again.
+            if (_unavailableErrorMessage != null)
+            {
+                return;
+            }
+
+            string tenantId = EnvironmentVariables.TenantId;
+            string clientId = EnvironmentVariables.ClientId;
+            string clientSecret = EnvironmentVariables.ClientSecret;
+            string username = EnvironmentVariables.Username;
+            string password = EnvironmentVariables.Password;
+            string sdkAuthLocation = EnvironmentVariables.SdkAuthLocation;
+
+            if (tenantId != null && clientId != null && ((username != null && password != null) || clientSecret != null))
+            {
+                if (clientSecret != null)
+                {
+                    _credential = new ClientSecretCredential(tenantId, clientId, clientSecret, _pipeline);
+                }
+                else
+                {
+                    _credential = new UsernamePasswordCredential(username, password, clientId, tenantId, _pipeline);
+                }
+            }
+            else if (sdkAuthLocation != null)
+            {
+                try
+                {
+                    _credential = BuildCredentialForCredentialsFile(isAsync ? await ParseCredentialsFileAsync(sdkAuthLocation, cancellationToken).ConfigureAwait(false) : ParseCredentialsFile(sdkAuthLocation));
+                }
+                catch (Exception e) when (!(e is OperationCanceledException))
+                {
+                    _unavailableErrorMessage = "Error parsing Azure SDK auth file";
+                    _unavailableException = e;
+                }
+            }
+            else
+            {
+                StringBuilder builder = new StringBuilder("Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Alternately, AZURE_AUTH_LOCATION may be used to specify the location of a Azure SDK Auth file.  Currently set variables [");
```

Comment:
I know that Jon has an open PR to basically remove all of this text.  It has definitely gotten longer.  I'm not sure if we'd rather reference some page in our documentation about how the EnvironmentCredential in the error (maybe via a shortlink)?